sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/SingletonPatternCheck.kt

```diff
@@ -0,0 +1,132 @@
+ */
+package org.sonarsource.kotlin.checks
+
+import org.jetbrains.kotlin.descriptors.ConstructorDescriptor
+import org.jetbrains.kotlin.psi.KtBlockExpression
+import org.jetbrains.kotlin.psi.KtCallExpression
+import org.jetbrains.kotlin.psi.KtClass
+import org.jetbrains.kotlin.psi.KtClassBody
+import org.jetbrains.kotlin.psi.KtFile
+import org.jetbrains.kotlin.psi.KtFunctionLiteral
+import org.jetbrains.kotlin.psi.KtLambdaArgument
+import org.jetbrains.kotlin.psi.KtLambdaExpression
+import org.jetbrains.kotlin.psi.KtModifierListOwner
+import org.jetbrains.kotlin.psi.KtObjectDeclaration
+import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.psi.KtPropertyDelegate
+import org.jetbrains.kotlin.psi.KtReferenceExpression
+import org.jetbrains.kotlin.psi.allConstructors
+import org.jetbrains.kotlin.psi.psiUtil.isPrivate
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.calls.util.getCall
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+import org.sonar.check.Rule
+import org.sonarsource.kotlin.api.AbstractCheck
+import org.sonarsource.kotlin.api.ConstructorMatcher
+import org.sonarsource.kotlin.plugin.KotlinFileContext
+import org.sonarsource.kotlin.visiting.KtTreeVisitor
+
+@Rule(key = "S6515")
+class SingletonPatternCheck : AbstractCheck() {
+
+    override fun visitKtFile(file: KtFile, kotlinFileContext: KotlinFileContext) {
+        val singletonClassCandidateExtractor = SingletonClassCandidateExtractor()
+        singletonClassCandidateExtractor.visitTree(file)
+        val singletonClassCandidates = singletonClassCandidateExtractor.singletonClassCandidates
+        if (singletonClassCandidates.isEmpty()) return
+
+        val singleConstructorCallExtractor =
+            SingleConstructorCallExtractor(singletonClassCandidates, kotlinFileContext.bindingContext)
+        singleConstructorCallExtractor.visitTree(file)
+        val singleConstructorCallByClass = singleConstructorCallExtractor.singleConstructorCallByClass
+        if (singleConstructorCallByClass.isEmpty()) return
+
+        singleConstructorCallByClass.values.stream().filter {
+            isInitializingCall(it) || isLazyInitializingCall(it, kotlinFileContext.bindingContext)
+        }.forEach {
+            kotlinFileContext.reportIssue(it, "Singleton pattern should use object declarations or expressions")
+        }
+    }
+}
+
+private class SingletonClassCandidateExtractor : KtTreeVisitor() {
+
+    val singletonClassCandidates = mutableSetOf<String>()
+
+    override fun visitClass(klass: KtClass) {
+        if (isSingletonClassCandidate(klass)) {
+            klass.fqName?.let {
+                singletonClassCandidates.add(it.toString())
+            }
+        }
+    }
+}
+
+private fun isSingletonClassCandidate(klass: KtClass): Boolean {
+    val constructors = klass.allConstructors
+    return klass.isPrivate() || (constructors.isNotEmpty() && constructors.all(KtModifierListOwner::isPrivate))
```

Comment:
I would probably write 
```
constructors.all { it.isPrivate() }
```
instead of passing a function reference, out of readability reasons.

However, I don't think we have a clearly defined rule here and this is my personal taste. Seeing this, maybe it's a good topic for a coffee break one day.

Reply:
Fixed, since I know now that the compiler generates the same code for both.